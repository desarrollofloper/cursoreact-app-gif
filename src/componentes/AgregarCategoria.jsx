import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faMagnifyingGlass } from '@fortawesome/free-solid-svg-icons';
import { useState } from 'react';

export const AgregarCategoria = ({ onAgregarCategoria /* setCategorias */ }) => {
    const [inputValue, setInputValue] = useState('');
    const handleInputValueChange = (event) => {
        setInputValue(event.target.value);
    };
    const handleSubmitform = (event) => {
        event.preventDefault();
        console.log(event);
        if(inputValue.trim().length < 2) return;
        onAgregarCategoria(inputValue, () => { setInputValue(''); }, (codigo = '') => {
            if(codigo == 'duplicado') setInputValue('');
        });
        // setCategorias(categorias => [inputValue, ...categorias]);
    };
    return (
        <form onSubmit={ handleSubmitform }>
            <div className='bloque-input mt-2'>
                <div style={{ display:'flex', alignItems:'center', }}>
                    <input type='search' placeholder='Buscar' className='mr-3 fs-1_4em' value={ inputValue } onChange={ handleInputValueChange } required />
                    <button type='submit' className='boton botonTema'>
                        <FontAwesomeIcon icon={ faMagnifyingGlass } />
                    </button>
                </div>
            </div>
        </form>
    )
}
