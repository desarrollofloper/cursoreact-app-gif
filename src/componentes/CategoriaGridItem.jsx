import PropTypes from 'prop-types';

export const CategoriaGridItem = ({gif}) => {
    return (
        <a className='CategoriaGridItem' href={ gif.url } target='_blank'>
            <img src={ gif.imagen_url } title={ gif.titulo } />
            { gif.titulo ? (
                <span className='detalleItem'>
                    <span>{ gif.titulo }</span>
                </span>
            ) : <></> }
        </a>
    )
}

CategoriaGridItem.propTypes = {
    gif: PropTypes.shape({
        // id: PropTypes.string,
        url: PropTypes.string.isRequired,
        imagen_url: PropTypes.string.isRequired,
        titulo: PropTypes.string,
        // imagen_original_url: PropTypes.string,
    }),
}
