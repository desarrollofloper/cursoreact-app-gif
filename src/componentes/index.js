export * from './AgregarCategoria';
export * from './CategoriaGrid';
export * from './CategoriaGridItem';