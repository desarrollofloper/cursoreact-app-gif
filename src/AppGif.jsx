import { useState } from 'react';
import { AgregarCategoria, CategoriaGrid } from './componentes';
import './assets/css/AppGif.css';

export const AppGif = () => {
    // const [categoriaBusqueda, setCategoriaBusqueda] = useState('');
    const [categorias, setCategorias] = useState([]);

    const handleAgregarCategoria = (categoria, onSuccess, onError) => {
        if(categoria.trim().length < 5){
            onError('busquedacorta');
            return;
        }
        if((categorias.filter(item => item.toLowerCase() == categoria.toLowerCase())).length){
            onError('duplicado');
            const categoriasReordenadas = categorias.filter(item => item.toLowerCase() != categoria.toLowerCase());
            categoriasReordenadas.unshift(categoria);
            setCategorias(categoriasReordenadas);
            return;
        }
        setCategorias([categoria, ...categorias]);
        onSuccess();
    };

    const bloqueCategoriaGrids = categorias.length ? (
        categorias.map(item => (
            <CategoriaGrid
                key={ item + '_' + (Math.random() * 1000000) }
                categoria={ item }
            />
        ))
    ) : ( <div>No has realizado alguna b√∫squeda.</div> );

    return (
        <>
            <header data-testid='header'>
                <div style={{ display: 'inline-flex', alignItems: 'center', }}>
                    <h1 style={{ marginRight: '1rem', }}>App Gif</h1>
                </div>
                <AgregarCategoria
                    onAgregarCategoria={ handleAgregarCategoria }
                />
            </header>
            <main data-testid='main'>
                <div>
                    { bloqueCategoriaGrids }
                </div>
            </main>
        </>
    );
}
