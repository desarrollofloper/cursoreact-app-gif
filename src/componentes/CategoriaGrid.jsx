// import PropTypes from 'prop-types';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSpinner } from '@fortawesome/free-solid-svg-icons';
import { CategoriaGridItem } from './CategoriaGridItem';
import { useObtenerGifs } from '../hooks/useObtenerGifs';

export const CategoriaGrid = ({ categoria }) => {
    const {gifs, cargandoGifs} = useObtenerGifs(categoria);
    const backgroundImage = !cargandoGifs && gifs.length && `url(${ gifs[0].images.original.url })`;
    console.log(gifs);
    return (
        <div className='CategoriaGrid' style={{ backgroundImage }}>
            <h2>{ categoria }</h2>
            <div className='mt-2'>
                {
                    cargandoGifs
                    ? <div>Cargando<FontAwesomeIcon icon={ faSpinner } spin className='ml-2' /></div>
                    : (
                        gifs.length
                        ? <div className="contenedorCategoriaGridItem">{ gifs.map(gif => ( <CategoriaGridItem key={ gif.id } gif={ gif } /> )) }</div>
                        : <div>Sin coincidencias</div>
                    )
                }
            </div>
        </div>
    )
}

// CategoriaGrid.propTypes = {
//     categoria: PropTypes.string.isRequired,
// }
