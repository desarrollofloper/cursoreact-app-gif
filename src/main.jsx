import React from 'react'
import ReactDOM from 'react-dom/client'
import { AppGif } from './AppGif';

ReactDOM.createRoot(document.getElementById('root')).render(
  <React.StrictMode>
    <AppGif />
  </React.StrictMode>,
)
