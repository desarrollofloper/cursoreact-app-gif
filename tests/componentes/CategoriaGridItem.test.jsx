import { render } from '@testing-library/react';
import { CategoriaGridItem } from '../../src/componentes';

describe('Pruebas en CategoriaGridItem', () => {
    test('Snapshot', () => {
        const gif = {
            url: '',
            imagen_url: '',
            titulo: '',
        };
        const { container } = render( <CategoriaGridItem gif={ gif } /> );
        expect(container).toMatchSnapshot();
    });
});