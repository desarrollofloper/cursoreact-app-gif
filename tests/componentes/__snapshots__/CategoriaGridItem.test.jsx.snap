// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en CategoriaGridItem Snapshot 1`] = `
<div>
  <a
    class="CategoriaGridItem"
    href=""
    target="_blank"
  >
    <img
      src=""
      title=""
    />
  </a>
</div>
`;
